pex/pex_builder.py

```diff
@@ -224,6 +226,69 @@ def set_script(self, script):
         'Could not find script %r in any distribution %s within PEX!' % (
             script, ', '.join(str(d) for d in self._distributions)))
 
+  def _verify_entry_point(self, entry_point):
+    """
+    Given entry_point `a.b.c:m`, make sure
+    1. a/b/c.py exists
+    2. `m` method is defined in a/b/c.py
+
+    :param entry_point:
+    :return:
+    """
+    ep_split = entry_point.split(':')
+    # Only module is specified
+    if len(ep_split) == 1:
+      ep_module = ep_split[0]
+      ep_method = None
+    elif len(ep_split) == 2:
+      ep_module = ep_split[0]
+      ep_method = ep_split[1]
+    else:
+      raise self.InvalidEntryPoint("Fail to parse: `{}`".format(entry_point))
+
+    # a.b.c:m ->
+    # ep_module_elements = ['a','b','c.py']
+    ep_module_elements = ep_module.split(os.sep)
+    ep_module_elements[-1] = ep_module_elements[-1] + '.py'
+
+    file_candidates = self._gather_file_candidates(ep_module_elements)
+    if not file_candidates:
+      raise self.InvalidEntryPoint("{} does not exist for entry point `{}`."
+                                   .format(os.path.sep.join(ep_module_elements), entry_point))
+
+    if len(file_candidates) > 1:
+      self._logger.warn("More than one file found for {}".format(ep_module))
+
+    if not ep_method:
+      return
+    # 2. Given files matching `a.b.c:m`, verify method `m` exists.
+    final_candidates = []
+    for c in file_candidates:
+      with open(os.path.join(self._chroot.path(), c), 'r') as f:
+        tree = ast.parse(f.read())
+        methods = filter(lambda x: isinstance(x, ast.FunctionDef) and x.name == ep_method, tree.body)
+        if methods:
+          final_candidates.append(c)
+
+    if not final_candidates:
+      raise self.InvalidEntryPoint("No method name `{}` has been found in {}."
+                                   .format(ep_method, file_candidates))
+
+  def _gather_file_candidates(self, entry_point_elements):
+    # 1. Find the files matching a.b.c:m, which is a/b/c.py
```

Comment:
Importing the code feels quite sketchy, even out of process, because they could have bad side effect or even security concerns.

How do you feel about putting verifying entry point behind an option, stating that it would only check for static method in
* `a/b/c.py`
* `a/b/c/__init__.py`
and allow people to skip static check by disabling it?

Reply:
If validation is opt-in, it seems like import would be the way to go. The user is going to run the pex after all and suffer the side-effects soon anyhow when they do so! Basically, this is forced to be a half measure one way or the other, so the choice comes down to avoiding side-effects at build time (they'll happen at runtime) for the X% of cases that have this pathology, or not supporting Y% of cases that have bizarre dynamic module generation. I'm really not sure what these percentages are or what is the better way to go. Import just seems attractive since it's robust: ie it's what pex does at runtime. Parsing is new logic.

Reply:
Yeah that's fair. Importing could also get around having to workaround ast parsing between python 2 and 3.

So I am going to do the importing, and then put it behind an option with default disabled.